### ISSUE TITLE
1.2.5+1143-0ppa does not start on Ubuntu 12.04.1 Precise 

### ISSUE THREAD

**Issue by amorphic** [mobile-shell/mosh#758]
#727 causes the 1.2.4a mosh package for Ubuntu Precise to fail.

This was fixed and released as 1.2.5. The [latest version for Precise in the mosh PPA is 1.2.5+1143-0](https://launchpad.net/~keithw/+archive/ubuntu/mosh-dev?field.series_filter=precise).

However this release also incorporated 389fc232051087f2430e9e7b05f770a38069cb39 from July 2015 for *Better, more flexible handling for IPv4/IPv6.*

Now the mosh perl wrapper script throws this error:
```
"addrinfo" is not defined in %Socket::EXPORT_TAGS at /usr/bin/mosh line 38
        main::BEGIN() called at /usr/bin/mosh line 38
        eval {...} called at /usr/bin/mosh line 38
Can't continue after import errors at /usr/bin/mosh line 38
BEGIN failed--compilation aborted at /usr/bin/mosh line 38.
```

I suspect the new IPv4/IPv6 handling is using features missing from `Socket 1.94` which is included with the latest system Perl for Precise `5.14.2-6`.

**Comment by cgull** [comment]
That fixes it for Ubuntu 12.04.  However, using getaddrinfo() in general breaks on Perl < 5.14.  This can be fixed too by bringing in `Socket::GetAddrInfo` for those older versions, and fixing up dependencies.  I've chosen not to do this right now.  Does anybody care?

@amorphic:  What problem was causing failures on 1.2.4a for you?  It can't be #727.

**Comment by keithw** [comment]
I'm okay with the dependency on Perl 5.14, but what is the failure mode when run on an earlier version? I'm sure there will be people on (insert random Unix) who will have an older version, and I'd rather they get a comprehensible error message than end up confused and come here or to IRC...

**Comment by cgull** [comment]
```shell
Perl v5.14.0 required--this is only v5.12.2, stopped at /home/cgull/src/mosh-keithw/scripts/mosh line 33.
BEGIN failed--compilation aborted at /home/cgull/src/mosh-keithw/scripts/mosh line 33.
```
Without the `use` statement, we get compile failures on the flag definitions (`AI_NUMERICADDR`, etc).

**Comment by cgull** [comment]
Actually, the errors in that case are
```
"getaddrinfo" is not exported by the Socket module
 "getnameinfo" is not exported by the Socket module
 "AI_CANONNAME" is not exported by the Socket module
 "AI_NUMERICHOST" is not exported by the Socket module
 "NI_NUMERICHOST" is not exported by the Socket module
Can't continue after import errors at /home/cgull/src/mosh-keithw/scripts/mosh line 38
BEGIN failed--compilation aborted at /home/cgull/src/mosh-keithw/scripts/mosh line 39.
```

**Comment by keithw** [comment]
The explicit error seems much better! I'm just worried about the support burden from people crawling out of the woodwork on HP-UX (or RHEL?) 2011 edition.

**Comment by amorphic** [comment]
@cgull re 1.2.4a - mosh was working fine for a long time, then we suddenly started experiencing the symptoms of [this bug](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=817929).

Also note that the system perl on the machine in question is `5.14.2-6ubuntu2.5` but we're still seeing this error.

**Comment by cgull** [comment]
@amorphic:

Which symptoms?  If you're talking about just the symptoms in the original report, that's unfortunately a rather generic report covering a lot of possible problems.  If you think the problem includes the specific issues discussed in the many comments on that bug, or is otherwise interesting, let us know.  Try `ssh -t <remotehost> mosh-server` for possibly-more-informative errors.

The problems in #727 are caused by a new issue in glibc-2.22 and later.  Ubuntu 12.04 has glibc-2.15.  So I don't see how you could be running into those issues, unless you're actually using glibc-2.22 or later.

**Comment by amorphic** [comment]
@cgull yep it's possible I jumped to conclusions. Basically mosh was running fine on the host in question for a long time and then suddenly users started getting the following when attempting to connect:

```
mosh did not make a successful connection to X.X.X.X:60101.
Please verify that UDP port 60101 is not firewalled and can reach the server.

(By default, mosh uses a UDP port between 60000 and 61000. The -p option
selects a specific UDP port number.)

[mosh is exiting.]
```

Around the same time the bug was reported and I assumed that was our problem.

With a fix for #759 now merged I might just wait for a new `1.2.5` release for Precise and see how that goes.
